onnx/shape_inference/implementation.h

```diff
@@ -71,8 +73,15 @@ struct InferenceContextImpl : public InferenceContext {
       const auto inputDataIter = inputDataByName.find(input);
       if (inputDataIter != inputDataByName.cend()) {
         allInputData_.push_back(inputDataIter->second);
+        allInputSparseData_.push_back(nullptr);
       } else {
         allInputData_.push_back(nullptr);
+        const auto inputSparseDataIter = inputSparseDataByName.find(input);
```

Comment:
This means typeAndShapeInferenceFunction will have to check the input type and based on the type call the right method of InferenceContext getInputData or getInputSparseData?

Reply:
> This means typeAndShapeInferenceFunction will have to check the input type and based on the type call the right method of InferenceContext getInputData or getInputSparseData?

I think, each kernel would actually query what is required. Currently, all kernels know they would be recieiving Tensors both as input and initializers. The upcoming kernels would query only Sparse Data. In general, it might be both.
The issue here and through out the code, is that we initially have two different fields, initializer and sparse_initializer. And Tensor and SparseTensor. Both of these tuples have the same high level semantic meaning but different types in proto and implementation.
So we are bound to have two different types and at times we would be checking both.